Mathlib/Analysis/MeanInequalities.lean

```diff
@@ -190,6 +190,48 @@ theorem geom_mean_eq_arith_mean_weighted_of_constant (w z : ι → ℝ) (x : ℝ
     ∏ i ∈ s, z i ^ w i = ∑ i ∈ s, w i * z i := by
   rw [geom_mean_weighted_of_constant, arith_mean_weighted_of_constant] <;> assumption
 
+
+
+/- **AM-GM inequality - equality condition**: This theorem provides the equality condition for the
+weighted version of the AM-GM inequality for real-valued nonnegative functions. Note that the w i
+here need to be positive.-/
+theorem geom_mean_arith_mean_weighted_eq_iff (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 < w i)
+    (hw' : ∑ i ∈ s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) :
+    ∏ i ∈ s, z i ^ w i = ∑ i ∈ s, w i * z i ↔ ∀ j ∈ s, z j = ∑ i ∈ s, w i * z i := by
```

Comment:
Can you adapt the proof to show this?
```suggestion
theorem geom_mean_arith_mean_weighted_eq_iff (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
    (hw' : ∑ i ∈ s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) :
    ∏ i ∈ s, z i ^ w i = ∑ i ∈ s, w i * z i ↔ ∀ j ∈ s, 0 < w j → z j = ∑ i ∈ s, w i * z i := by
```

Reply:
Sure, I’d be happy to do that.

Reply:
I noticed that in a similar lemma `strictConvexOn_exp.map_sum_eq_iff'`, it uses `w j ≠ 0 →` rather than `0 < w j →`. Is it better to use `≠` or `<` here?

```
lemma StrictConvexOn.map_sum_eq_iff' (hf : StrictConvexOn 𝕜 s f) (h₀ : ∀ i ∈ t, 0 ≤ w i)
    (h₁ : ∑ i ∈ t, w i = 1) (hmem : ∀ i ∈ t, p i ∈ s) :
    f (∑ i ∈ t, w i • p i) = ∑ i ∈ t, w i • f (p i) ↔
      ∀ j ∈ t, w j ≠ 0 → p j = ∑ i ∈ t, w i • p i
```

Reply:
I wrote two more version based on the first version, `geom_mean_arith_mean_weighted_eq_iff'` and `geom_mean_arith_mean_weighted_eq_iff''`.
But the linter said something like "missing a doc-string".
I'm not sure how to properly document this -- any guidance? Thanks.

```
theorem geom_mean_arith_mean_weighted_eq_iff' (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
    (hw' : ∑ i ∈ s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) :
    ∏ i ∈ s, z i ^ w i = ∑ i ∈ s, w i * z i ↔ ∀ j ∈ s, w j ≠ 0 → z j = ∑ i ∈ s, w i * z i := by
  have h (i) (_ : i ∈ s) : w i * z i ≠ 0 → w i ≠ 0 := by aesop
  have h' (i) (_ : i ∈ s) : z i ^ w i ≠ 1 → w i ≠ 0 := by aesop
  rw [← sum_filter_of_ne h, ← prod_filter_of_ne h', geom_mean_arith_mean_weighted_eq_iff]
  · simp
  · simp (config := { contextual := true }) [(hw _ _).gt_iff_ne]
  · rwa [sum_filter_ne_zero]
  · simp_all only [ne_eq, mul_eq_zero, not_or, not_false_eq_true, and_imp, implies_true, mem_filter]

theorem geom_mean_arith_mean_weighted_eq_iff'' (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
    (hw' : ∑ i ∈ s, w i = 1) (hz : ∀ i ∈ s, 0 ≤ z i) :
    ∏ i ∈ s, z i ^ w i = ∑ i ∈ s, w i * z i ↔ ∀ j ∈ s, 0 < w j → z j = ∑ i ∈ s, w i * z i := by
  have h (i) (a : i ∈ s) : w i * z i ≠ 0 → 0 < w i := fun _ => lt_of_le_of_ne' (hw i a) (by aesop)
  have h' (i) (a : i ∈ s) : z i ^ w i ≠ 1 → 0 < w i := fun _ => lt_of_le_of_ne' (hw i a) (by aesop)
  rw [← sum_filter_of_ne h, ← prod_filter_of_ne h', geom_mean_arith_mean_weighted_eq_iff]
  · simp
  · simp (config := { contextual := true }) [(hw _ _).gt_iff_ne]
  · have : filter (fun x => 0 < w x) s = filter (fun x => w x ≠ 0) s := by
      apply filter_congr
      exact fun x hx => ⟨fun a => (ne_of_lt a).symm, fun a => lt_of_le_of_ne (hw x hx) a.symm⟩
    rwa [this, sum_filter_ne_zero]
  · simp_all only [ne_eq, mul_eq_zero, not_or, not_false_eq_true, and_imp, implies_true, mem_filter]
```

![image](https://github.com/user-attachments/assets/56d679f4-6656-463d-8f86-3ff0abed4f25)

Reply:
`≠` is better, sorry for the misguidance